### PR TITLE
utilities for async dependency injection (GDEV-845)

### PR DESCRIPTION
**Pull request by KerstenBreuer**
```
- Added a class that extends the `dependency_injector`'s Resource provider
   to support classes with a `construct` method that returns an async
   context manager
- Added a class adds an async context manager interface to the
   `DeclarativeContainer` of the `dependency_injector` framework to handle
   the setup and teardown of resources
- Added a helper function that automatically chooses the correct constructor
   for a given class
````

### CODE REVIEW

File: tests/unit/test_inject.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    constructable: ContextConstructable, exception: Optional[Exception]
+):
+    """Test whether init and shutdown correctly works with a context constructor."""
+
+    foo = "bar"
+
+    test = ContextConstructor(constructable, foo)
+
+    with pytest.raises(exception) if exception else nullcontext():  # type: ignore
+        await test()
```
Comment (Cito) [inline tests/unit/test_inject.py:108]:
```suggestion
        resource = await test.async_()
        assert isinstance(resource, constructable)
```

File: hexkit/inject.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+import inspect
+from typing import Any, AsyncContextManager, AsyncIterator, Callable, Optional, TypeVar
+
+import dependency_injector.containers
+import dependency_injector.providers
+
+from hexkit.custom_types import ContextConstructable
+
+
```
Comment (Cito) [inline hexkit/inject.py:35]: You should add the things that his module exports for usage here:

```python
__all__ = [
    'get_constructor', 'ContainerBase',
    'NotConstructableError', 'AsyncInitShutdownError'
]
```

File: hexkit/inject.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+Utilities for dependency injection based on the dependency_injector framework.
+
```
Comment (Cito) [inline hexkit/inject.py:19]: You should state the main purpose of the module here, something like:

"This module contains a helper function that automatically selects the suitable "Provider" class for the dependency_injector framework and makes the "Container" class usable as async context manager."

**Review by Cito — APPROVED** [review]
Looks good, I think that should do it for now.

I think it's important for hexkit to have good examples and docs, so that people don't need to read the source and understand all aspects of the dependency_injector framework.

File: hexkit/inject.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    async context manager.
+    """
+
+    if not (
+        hasattr(constructable, "construct")
+        # Using isinstance together with Callable should be supported,
+        # see: https://peps.python.org/pep-0484/#callable
+        # Thus ignoring mypy error:
+        and isinstance(constructable.construct, Callable)  # type: ignore
+    ):
```
[the remark this answers is not in the document]
Comment (KerstenBreuer) [reply hexkit/inject.py:60]: Nice

File: hexkit/inject.py (hunk 2 above)
Comment (KerstenBreuer) [reply hexkit/inject.py:35]: Thanks, good point

Comment (KerstenBreuer) [reply hexkit/inject.py:35]: Done

File: hexkit/inject.py (hunk 3 above)
Comment (KerstenBreuer) [reply hexkit/inject.py:21]: Done.

**Comment by KerstenBreuer** [comment]
Thanks a lot @Cito for the nice review.

**Pull request merged**